### PR TITLE
stats/opentelemetry: CSM Observability server side component changes

### PR DESCRIPTION
**Pull request by zasweq**
RELEASE NOTES: N/A

### CODE REVIEW

File: stats/opentelemetry/server_metrics.go [hunk 1]
```diff
@@ -105,6 +218,9 @@ func (ssh *serverStatsHandler) HandleRPC(ctx context.Context, rs stats.RPCStats)
 func (ssh *serverStatsHandler) processRPCData(ctx context.Context, s stats.RPCStats, ai *attemptInfo) {
 	switch st := s.(type) {
 	case *stats.InHeader:
+		if ai.pluginOptionLabels == nil && ssh.o.MetricsOptions.pluginOption != nil {
+			ai.pluginOptionLabels = ssh.o.MetricsOptions.pluginOption.GetLabels(st.Header)
```
[the remark this answers is not in the document]
Comment (zasweq) [reply stats/opentelemetry/server_metrics.go]: Ah, right. Switched.

File: stats/opentelemetry/server_metrics.go [hunk 2]
```diff
@@ -55,6 +56,118 @@ func (ssh *serverStatsHandler) initializeMetrics() {
...
+		md.Append("x-envoy-peer-metadata", val...)
+	}
+
+	return als.ServerStream.SetHeader(md)
+}
+
+func (als *attachLabelsStream) SendHeader(md metadata.MD) error {
+	if !als.attachedLabels.Swap(true) {
+		val := als.metadataExchangeLabels.Get("x-envoy-peer-metadata")
+		md.Append("x-envoy-peer-metadata", val...)
```
[the remark this answers is not in the document]
Comment (zasweq) [reply stats/opentelemetry/server_metrics.go]: Done.

File: stats/opentelemetry/server_metrics.go [hunk 3]
```diff
@@ -55,6 +56,118 @@ func (ssh *serverStatsHandler) initializeMetrics() {
...
+func (alts *attachLabelsTransportStream) SendHeader(md metadata.MD) error {
+	if !alts.attachedLabels.Swap(true) {
+		val := alts.metadataExchangeLabels.Get("x-envoy-peer-metadata")
+		md.Append("x-envoy-peer-metadata", val...)
+	}
+
+	return alts.ServerTransportStream.SendHeader(md)
+}
+
+func (ssh *serverStatsHandler) unaryInterceptor(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
```
[the remark this answers is not in the document]
Comment (dfawley) [reply stats/opentelemetry/server_metrics.go:83]: I'm fine to not change existing code, but if you want to, feel free and I will approve.

File: stats/opentelemetry/server_metrics.go [hunk 4]
```diff
@@ -49,10 +50,117 @@ func (ssh *serverStatsHandler) initializeMetrics() {
...
+	if ssh.o.MetricsOptions.pluginOption != nil {
+		metadataExchangeLabels = ssh.o.MetricsOptions.pluginOption.GetMetadata()
+	}
+	als := &attachLabelsStream{
+		ServerStream:           ss,
+		metadataExchangeLabels: metadataExchangeLabels,
+	}
+	err := handler(srv, als)
+	if err != nil {
+		logger.Infof("RPC failed with error: %v", err)
```
Comment (dfawley) [inline stats/opentelemetry/server_metrics.go:155]: Remove this log, too, please.

I don't see why we would want to log when an RPC fails in the otel plugin.  It's not really a significant event, and could become extremely spammy.  If you need it for debugging then you can `V(2)` it.

https://github.com/grpc/grpc-go/blob/master/Documentation/log_levels.md#info

> At verbosity of 0 (the default), any single info message should not be output more than once every 5 minutes under normal operation.

File: stats/opentelemetry/server_metrics.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply stats/opentelemetry/server_metrics.go:83]: Alright I'll leave as is.

File: stats/opentelemetry/server_metrics.go (hunk 4 above)
Comment (zasweq) [reply stats/opentelemetry/server_metrics.go:155]: Ok I had this since every? most? interceptors I saw in the code base had this.

Comment (dfawley) [reply stats/opentelemetry/server_metrics.go]: What interceptors are those?

File: stats/opentelemetry/server_metrics.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply stats/opentelemetry/server_metrics.go:83]: Done.

**Review by dfawley — APPROVED** [review]

File: stats/opentelemetry/server_metrics.go (hunk 4 above)
Comment (dfawley) [reply stats/opentelemetry/server_metrics.go]: Examples are not library code, though, so that is fine.

**Pull request merged**